design/secure-communication-design.md

```diff
@@ -0,0 +1,88 @@
+# Secure Communication between Contour and Envoy
+
+_Status_: Draft, in review.
+
+This document outlines how to secure the communication between Contour
+and Envoy containers.
+
+## Goals
+
+- Achieve secure communication between Contour and Envoy
+- Authenticate Envoy connecting to Contour
+- Rotating and invalidating the credentials
+
+## Non-goals
+
+- Support for dynamic credentials
+- Support for use of Secret Discovery Service for credential management
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Default options for secure communication
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. The default `contour bootstrap` will generate the bootstrap configuration that
+enables secure communication between the containers. If `contour bootstrap` is passed with
+the parameter `--disable-tls`, the bootstrap configuration generated will make use of insecure
+communication between the containers.
+
+### Bootstrapping process
+
+This bootstrap phase is independent of `contour bootstrap` and uses `contour bootstrap-tls`.
+`contour bootstrap-tls` is run as an init container to bootstrap the TLS configuration.
+The bootstrap phase will create three pairs of the TLS certificates. The details about
+each pair of certificates are mentioned below:
+1. The first pair is used by Contour to serve HTTPS.
```

Comment:
I'm not sure generating certs that envoy uses to serve TLS is in the scope of this issue. Those should be part of an `Ingress` or `IngressRoute` object.

Reply:
I am not sure if I understand your review comment here. Let me try to drive my idea here. It is to serve Contour the xDS APIs over HTTPS and Envoy talks to Contour over mTLS. To setup the mTLS communication, I am proposing to create a CA cert and get two pairs of the certs signed with the CA cert. One pair of the cert is used by Contour to serve HTTPS and Envoy uses the other pair to connect to Contour over mTLS.

Reply:
You are 100% correct, please ignore my comment, I forgot this was mTLS and not just TLS.